### PR TITLE
Towards new BlockChain

### PR DESCRIPTION
**Pull request by nearmax**
No-op change.

* Introduced wrapper around column storage and cache that achieves:
    * It does not copy the object on the lookup. Previous implementation would copy the object even if it is in cache;
    * Allows beacon chain and shard chains to use independent sets of columns. Previously beacon chain and shard chain would reuse each other columns which is not parallelizable and led to a bug where beacon chain and shard chain would overwrite each other's index;
   * Forward compatible usage of storage. If we add columns or shads later we will be able to reuse old storage files;
   * Recoverable failures on storage read/write. Recoverable failures make sure the cache does not go out of sync with the disk storage. Previously these were hard failures.

* Removed StateDb and replaced its usage with Trie, because StateDb was forwarding almost everything to Trie. StateDbUpdate is renamed to TrieUpdate.
* In order to break cyclic dependencies (that are getting more complex) moved types from beacon, shard, and chain crates into primitives. Also temporarily merged transaction crate into primitives.

### CODE REVIEW

File: core/storage/src/trie/mod.rs [hunk 1]
```diff
@@ -181,8 +184,8 @@ pub struct Trie {
 pub type DBChanges = HashMap<Vec<u8>, Option<Vec<u8>>>;
 
 impl Trie {
-    pub fn new(storage: Arc<KeyValueDB>, column: Option<u32>) -> Self {
-        Trie { storage, column, null_node: Trie::empty_root() }
+    pub fn new(storage: Arc<KeyValueDB>) -> Self {
+        Trie { storage, column: COL_STATE, null_node: Trie::empty_root() }
```
Comment (ilblackdragon) [inline core/storage/src/trie/mod.rs]: what if we need trie on a different column?

File: core/storage/src/trie/update.rs [hunk 2]
```diff
@@ -0,0 +1,348 @@
...
+        }
+    }
+    pub fn set(&mut self, key: &[u8], value: &DBValue) {
+        self.prospective.insert(key.to_vec(), Some(value.to_vec()));
+    }
+    pub fn remove(&mut self, key: &[u8]) {
+        self.prospective.insert(key.to_vec(), None);
+    }
+    pub fn for_keys_with_prefix<F: FnMut(&[u8])>(&self, prefix: &[u8], mut f: F) {
+        // TODO: join with iterating over committed / perspective overlay here.
```
Comment (ilblackdragon) [inline core/storage/src/trie/update.rs]: Add TODO(574) here please

File: node/beacon/src/types.rs [hunk 3]
```diff
@@ -1,112 +1,12 @@
 use parking_lot::RwLock;
```
Comment (ilblackdragon) [inline node/beacon/src/types.rs]: not really types anymore

File: core/storage/src/trie/mod.rs (hunk 1 above)
Comment (nearmax) [reply core/storage/src/trie/mod.rs]: It is WIP, the column that the Trie operates with will be hidden from the trie.

**Comment by ilblackdragon** [comment]
Filed comments under #575

File: core/storage/src/trie/update.rs (hunk 2 above)
Comment (nearmax) [reply core/storage/src/trie/update.rs]: Done

File: node/beacon/src/types.rs (hunk 3 above)
Comment (nearmax) [reply node/beacon/src/types.rs]: Done.

**Pull request merged**